core/hyphenator-liang.lua

```diff
@@ -95,11 +95,19 @@ SILE.hyphenator = {}
 SILE.hyphenator.languages = {}
 SILE._hyphenators = {}
 
+local function defaultHyphenateSegments (node, segments, _)
+  local hyphen = SILE.shaper:createNnodes(SILE.settings:get("font.hyphenchar"), node.options)
+  return SILE.nodefactory.discretionary({ prebreak = hyphen }), segments
```

Comment:
We the same createNnodes (invoking harbuzz) for discretionary alot = each time we are given bits of text from an unshaped segment...

I wonder whether we shouldn't font-cache them and measure performances.

Reply:
Yes I thought about that. I did some benchmarks with caching these per run of `hypenateNode()`, but couldn't see any improvement and due to variable scoping it wasn't going to make sense to try anything larger. If we did memoization sitting in front of a variant of `createNnodes()` (or even inside `createNnodes()` that could potentially be a big deal for longer documents.

Reply:
Oh also I didn't catch that this comment was on the *default* function. I did already test this one and couldn't come up with a measurable difference even in relatively long documents. The CA and TR ones are even less efficient right now, but I wasn't coming up with anything that was a measurable win as far as generating nodes once and reusing them. Not that it can't be done, I just didn't hit on any measurable bottleneck in the directly changed lines.

Reply:
I added memoization to the default hyphenator with a couple approaches. All 3 approaches I tried with various trade-offs for lifetimes and aggressiveness gave about the same performance: a 6% boost over not having anything. I couldn't make out any differences in speed between them so I'm going with by far the simplest code version. It only actually does any work inside segments where the options table ID remains the same, but that seems to be enough. The alternative is to take the time to serialize the options table content into something that is reusable as a key over a longer lifetime, but the cost of serializing the table seems to offset any gains in reusing results almost exactly.

Reply:
And never-mind. My benchmark test (~100 pages of Moby Dick) didn't hit it, but quite a few of our regression tests did. It will take some more careful handling across font and chunk changes to not have memoization be a disaster. That being said there is definitely some speed on the table here.

The speed isn't *lost* speed by this PR it is just something that's been sitting there all along, so I'll keep that branch separate and move on with this.